src/Data/Text/Internal/Encoding.hs

```diff
@@ -0,0 +1,522 @@
+{-# LANGUAGE BangPatterns, CPP, GeneralizedNewtypeDeriving, MagicHash,
+    UnliftedFFITypes #-}
+{-# LANGUAGE Trustworthy #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- |
+-- Internals of "Data.Text.Encoding".
+--
+-- If you'd like to depend on something from here that's not in "Data.Text.Encoding",
+-- please request to have it exported!
+module Data.Text.Internal.Encoding where
+
+#if defined(ASSERTS)
+import Control.Exception (assert)
+#endif
+import Control.Monad.ST (ST, runST)
+import Control.Monad.ST.Unsafe (unsafeIOToST, unsafeSTToIO)
+import Data.Bits ((.&.), shiftL, shiftR)
+import Data.ByteString (ByteString)
+import Data.Functor (void)
+import Data.Semigroup (Semigroup(..))
```

Comment:
This is a redundant import on newer GHCs. There are also redundant imports below.

Reply:
Should I hide it using CPP? `Semigroup` is needed for base 4.10 which is currently the lower bound for text.

Reply:
Yes, unfortunately `CPP` is the way, otherwise there is a flood of warnings from CI.